### PR TITLE
Implement `Spanned` to retrieve source locations on AST nodes

### PR DESCRIPTION
**Pull request by Nyrox**
This PR adds a new trait `Spanned` to retrieve source spans on AST nodes, see #161 , by recursively traversing the AST and combining spans. This approach is in contrast to the one taken in #839 and #790 by trying to minimise the amount of breaking changes for downstream users, by avoiding wrapping everything in `WithSpan<T>`.

## Main Changes

- Spanned trait and Span type added
- TokenWithLocation now stores Span, instead of Location
- Ident now stores a source span (omitted from Partial/-EQ for compatibility and tests)
- Some AST nodes store source tokens where not intrusive
- Parser::parse_keyword_token added

The general philosophy of the PR is to be "good enough" without breaking things. As a result certain expressions will have broken or incorrect spans. I imagine these can be cleaned up in future PRs (which might require breaking changes).

f.e. many expressions do not include keywords in their span i.e.
```sql
<expr> IS NOT NULL
|....|
```  
  
will have it's source span simply reported as `<expr>::span` and there is many such cases, some of which are easier to fix than others. For expressions we can't generate spans for, I use `Span::EMPTY` as a sort of sentinel value to indicate missing information.

With this approach the only downstream changes a user should have to do to upgrade, should be adding additional fields when matching on AST nodes.

## Future Work

- Store spans for `ast::value::Value`. This seems like a breaking change to me, which would require a `WithSpan<T>` like type
- Store keyword TokenWithLocation for expressions that currently don't have them
- Implement spans for the rest of the AST, namely Statements. I focused on getting Queries done.

### CODE REVIEW

**Comment by Nyrox** [comment]
@alamb Hey, we've started using this functionality internally with pretty great success so far. It's still a draft for now, because of the missing todo!'s, but I would appreciate feedback on the overall design and if you think this can get merged in the foreseeable future once issues are addressed. 😄

**Comment by alamb** [comment]
Thanks @Nyrox  -- I'll try and take a look over the next few days

cc @lovasoa @iffyio and @jmhain

File: src/parser/mod.rs
Comment (lovasoa) [inline src/parser/mod.rs:3206]: Wouldn't it be easier (and less breaking) to keep the existing `parse_keyword` function that returns a bool and add a new `get_keyword` or `parse_keyword_token` or `parse_keyword_with_loc` that returns a token with location ?

Comment (Nyrox) [reply src/parser/mod.rs:3206]: That's fine with me, I just figured no reason to have two functions that do the same thing, but one just omits information. And going forward I expect more and more places in the parser to want the Token version (error handling as well as more source spans). I don't expect any downstream users to be using internal parser methods, but I might be wrong. 😅 Hyrum's Law and all

**Comment by yuyang-ok** [comment]
How is this `PR` going? is it going to merge or what?

**Comment by Nyrox** [comment]
Alright, I have now gotten rid of all the todo!s and warnings and un-drafted the PR. There is a lot of missing implementations of spans and I have documented those (in hindsight maybe a derive macro would have been the way to go here, someone better at writing those than me is free to take a shot at it 😅 ).

File: src/parser/mod.rs
Comment (Nyrox) [reply src/parser/mod.rs]: I reverted the change and added `parse_keyword_token` as via your suggestion. I also changed `parse_keyword` to be implemented as `parse_keyword_token(keyword).is_some()` to not duplicate code. 😃

**Review by iffyio — COMMENTED** [review]
Thanks for tackling this @Nyrox! took a quick look and left some comments inline, I'll make some time to do another pass

File: src/ast/mod.rs
Comment (Nyrox) [reply src/ast/mod.rs:201]: Removing it sounds good to me, I just wanted to have the same behaviour as `Ident::with_quote` which still has the check, so it should prob. be removed in both places then

File: src/ast/spans.rs
Comment (Nyrox) [reply src/ast/spans.rs]: The intent here was to assert it doesn't panic 😅 I have replaced the old tests with slightly better ones

File: src/ast/spans.rs
Comment (Nyrox) [reply src/ast/spans.rs]: Smth. like that. I replaced it with smarter tests

File: src/ast/spans.rs
Comment (iffyio) [inline src/ast/spans.rs:1532]:
```suggestion
            .union_opt(&self.with_fill.as_ref().map(|f| f.span()))
```

File: src/ast/spans.rs
Comment (iffyio) [inline src/ast/spans.rs:1205]:
```suggestion
                .into_iter()
                .flatten()
```

**Review by iffyio — COMMENTED** [review]
Thanks @Nyrox! The changes look reasonable to me overall given the discussion in the GH issue. Left some comments, one mostly wondering around the equality behavior now that the token location is embedded within the AST

File: tests/sqlparser_common.rs
Comment (Nyrox) [reply tests/sqlparser_common.rs:9856]: Yeah I agree it can be a bit tricky. For this particular case it doesn't matter I think, because TokenWithLocation's custom Eq implementation ignoring spans. So as long as selects actually have _a_ select token they will compare.
What actually is the use case of having Eq on ast nodes? I guess it's so you can throw them into HashMaps?

File: src/ast/mod.rs
Comment (Nyrox) [reply src/ast/mod.rs]: 👍 good idea, done

File: tests/sqlparser_common.rs
Comment (iffyio) [reply tests/sqlparser_common.rs:9856]: Ah, no my mistake I missed that the span is being ignored in the comparison here as well. the actual token isn't currently ignored however? thinking that potentially makes it awkward needing to figure out what a token on a property should be, and not sure if there are guarantees provided by the parser on its value? (I'm assuming having the token included in the node is useful)

Not only Eq, but PartialEq as well, basically thinking as with today, any two subexpressions should remain equal regardless of what part of the tree they show up in (for example required when taking fingerprints of sql and comparing them)

Comment (Nyrox) [reply tests/sqlparser_common.rs:9856]: Hmm yeah. I mean "Token" in most of these cases is going to be Keyword, so there is some certainty to them being predictable. I.e. if two queries have different tokens, I would very much expect them to actually be different queries.

But I also get if it opens too many cans of worms, so we could just omit it from the trait implementations again, it's just a bit annoying then that we would have to start writing custom implementations for every type that is storing a Token.

The Token is useful, because you need to store it (or a at least it's Span) so you can include it for reported spans. 
I.e. `select a from b`, if you don't store the token for 'select' can't include that as part of it's span :/

File: src/ast/spans.rs
Comment (Nyrox) [reply src/ast/spans.rs:1230]: Good point, I'll go over the ones that are not doing this yet and fix them up 👍

File: src/ast/mod.rs
Comment (Nyrox) [reply src/ast/mod.rs:215]: What's the verdict here? Remove or not? :P I am not sure it's in the scope of this PR tbh.

File: src/ast/spans.rs
Comment (Nyrox) [reply src/ast/spans.rs]: Done. Thanks for the suggestion, I'll definitely stash this technique in my memory. 😄

File: src/ast/mod.rs
Comment (iffyio) [reply src/ast/mod.rs:215]: Ah I didn't realise the behavior exists already, it sounds reasonable to follow suit. The current behavior feels unusual to me though, and I agree that it'd be out of scope for this PR to change it. I can take a closer look at it and we can have the behavior here match the outcome

File: tests/sqlparser_common.rs
Comment (iffyio) [reply tests/sqlparser_common.rs:9856]: Yeah I agree it can become unweildy to try to workaround this by needing custom impls for each type, also as you mention its probably a reasonable assumption that the queries are different if they have different tokens (given that they come from the same parser)

File: src/ast/spans.rs
Comment (lovasoa) [inline src/ast/spans.rs:53]: @Nyrox : `for_clause` is specific to microsoft sql server (mssql), not mysql :)

Comment (Nyrox) [reply src/ast/spans.rs]: oops 😅 , fixed

**Comment by Nyrox** [comment]
Bump 😅

**Comment by lustefaniak** [comment]
Hi @Nyrox!
Thanks for this PR and pushing locations support further! 

We had an internal discussion to try to upstream the location changes we use in SYNQ. I found some time to finish rebase today and then noticed your PR. (https://github.com/apache/datafusion-sqlparser-rs/pull/1480)

I like your suggested changes, as they are very noninvasive for the downstream consumers. If this could be included in the main, that would be absolutely amazing. Having location inside every Ident is probably best. In my PR, I wrapped all Idents with location, but our fork uses them only when we need them. Having it always available inside of `Indent` sounds much better. 

It would also be good to design and include wrapping of nodes. It is essential, especially for enum cases where adding location to each of the enum cases would be too verbose. Also in some places those "smaller" locations just don't make much sense. In my PR, I'm using https://github.com/apache/datafusion-sqlparser-rs/pull/1480/commits/fc8e44afd66c3a5100a6546d8edba36c65870315#diff-fc6a8d66b8cb6bd48a119a70e489cbcef82e7f551e7cf08e8e972ef4e774ef49R12259-R12308 to align captured span to a non-whitespace portion of the query based on token indexes.

**Comment by Nyrox** [comment]
Hi @lustefaniak, thanks I remember seeing your fork when initially researching this 😅 
I agree something like a `WithSpan<T>` type will probably be needed to improve spans, especially in the case of enums as you said. I think for now the focus, as by the discussions with @alamb was mostly on just reducing the breaking surface as much as we can and get _something_ in. Maybe in a future PR we can start thinking about wrapping all the enums, but then at least we have a gradual roll out 😃

**Comment by Nyrox** [comment]
Hey all, think we've gotten quite far here - is there anything actively blocking this from getting approved? Are there any actions I can take to help you guys or do you just need the time to get to this item? 😊

**Comment by iffyio** [comment]
@Nyrox sorry for the delay in getting to this

I don't have any major comments regarding the impl and it seems to align with previous discussions so I'm suspecting this is close.

One aspect I'm not super clear on is how accurate/comprehensive we'll look to be regarding this feature going forward (primarily as it pertains to testing affected code paths), i.e. whether as a best effort, or similar to the other parser functionality (which expectedly requires more effort from PR authors when adding/updating features in the codebase) in order to properly power use cases like @lustefaniak mentions in #1480, or somewhere in-between.

In any case, I would wait for @alamb for when time allows to take a look and provide any thoughts/feedback on the overall approach!

**Comment by alamb** [comment]
Thanks @Nyrox and @iffyio -- i will try and find time this weke to review this propsoal

**Comment by mkarbo** [comment]
@alamb that's really appreciated 🙌

File: src/ast/mod.rs
Comment (Nyrox) [reply src/ast/mod.rs:223]: Can I close this comment, if this is gonna be done in a future PR or should we scram it in here if a decision is made? 😄

Comment (iffyio) [reply src/ast/mod.rs:223]: Yeah let's ignore is for this PR

**Comment by alamb** [comment]
Sorry for my silence / inability to find time to review this PR in depth. I am struggling with too many thing

TLDR I I think we should:
1. Get the tests passing on this PR and merge it in (if not all AST nodes have spans initially, we can file tickets to add the information over the next few releases)
2. Add some documentation explaining what is happening and how downstream consumers need to change (I think basically by ignoring the newly added fields

@iffyio  and I spoke about this a bit -- basically in my opinion adding source locations to sqlparser-rs is one of the last major remaining features it is missing.

However, it also has the potential to be very disruptive to downstream consumers.

It seems to me the requirements for this change are:
1. Any changes required by users of this crate should be mechanical
3. We should clearly document what changes are needed on upgrade (ideally with examples)

An example of a mechanical change might be adding
```rust
match node {  
  ast::Query { 
    field1,
    field2, 
    location: _,  // add a new line to ignored location
}
```

An example of a non mechanical change might be wrapping all AST nodes in some other struct so the structure of the `match` no longer works

I looked at this PR and I think it basically satisfies 1 for me so I think we should proceed. Would it be possible to write up something, perhaps in the `README` that explains  that we are adding new spans to the code and how we will handle them?

**Comment by Nyrox** [comment]
Okay, I wrote some documentation and fixed up the tests so CI should pass now (I hope :P )
Let me know what you think of the docs, I wasn't quite sure what to write so I hope it's decent.

@iffyio I ended up running into the problem we discussed further up about comparing queries. It slipped by me that Tokens store the raw string in addition to the Keyword enum, causing `select 5` and `SELECT 5` to compare as unequal. I have added a new wrapper type `IgnoreField` to work around this issue and make sure the tokens actually do not impact semantics. I am not exactly sure if this is _the best_ fix, but I think if we want queries to compare as semantically equal rather than syntactically then this is a decent solution for now. lmk what you think

File: src/ast/query.rs
Comment (iffyio) [inline src/ast/query.rs:284]: For the comment we can probably be more descriptive to mention that this points to the position of the SELECT keyword? (similar for others probably if we want to make it more obvious which part of the statement is being referred to)

File: src/ast/helpers/ignore_field.rs
Comment (iffyio) [inline src/ast/helpers/ignore_field.rs:23]: Instead of using a generic type would it make sense to represent via a regular wrapper type (thinking since the plan is only to wrap the tokens we attach to the nodes)? e.g.
```rust
pub struct AttachedToken(TokenWithLocation);

impl From<Token> for AttachedToken {}
impl From<TokenWithLocation> for AttachedToken {}
impl Hash for AttachedToken {}
// maybe even a default/noop impl to make it easier in scenarios like test cases where the info is irrelevant
// etc.

pub struct With {
    pub with_token: AttachedToken,
    pub recursive: bool,
    // ...
}
```

File: src/ast/query.rs
Comment (iffyio) [inline src/ast/query.rs:566]: can we update the comment to rather say what's being pointed to (thinking all attached tokens are implicitly needed for span reporting so that no need to treat this case specially)?

File: src/ast/mod.rs
Comment (Nyrox) [reply src/ast/mod.rs]: yup 👍

File: src/ast/query.rs
Comment (Nyrox) [reply src/ast/query.rs]: yup 👍

File: src/ast/query.rs
Comment (Nyrox) [reply src/ast/query.rs]: yup 👍

File: src/ast/helpers/ignore_field.rs
Comment (Nyrox) [reply src/ast/helpers/ignore_field.rs]: doneso 👍

File: src/ast/mod.rs
Comment (iffyio) [inline src/ast/mod.rs:970]:
```suggestion
    Wildcard(AttachedToken),
```

File: src/ast/mod.rs
Comment (iffyio) [inline src/ast/mod.rs:973]:
```suggestion
    QualifiedWildcard(ObjectName, AttachedToken),
```

File: src/ast/helpers/attached_token.rs
Comment (iffyio) [inline src/ast/helpers/attached_token.rs:27]: Ah I think we might  be missing the serde and visit derive on this struct given its attached to structs in the ast?

File: tests/sqlparser_postgres.rs
Comment (iffyio) [inline tests/sqlparser_postgres.rs:2523]: To make places like tests that don't care about the token leaner, could we add a helper function like `AttachedToken::empty()` that returns a dummy token?

File: src/tokenizer.rs
Comment (iffyio) [inline src/tokenizer.rs:572]: Ah do we still need these custom impls now that we have the wrapper type or we can go back to the derives for TokenWithLocation now?

Comment (Nyrox) [reply src/tokenizer.rs]: Yeah I think it's fine. Idk. if rust-analyzer can accurately do find references on trait implementations, but it did not find any, so I think it's only used upstream where we should have the wrapper now

File: tests/sqlparser_postgres.rs
Comment (Nyrox) [reply tests/sqlparser_postgres.rs]: I was sceptical about this, but it actually works pretty well 😅 
Not sure what Token to pick as a dummy token, so I went with `Token::EOF`

**Comment by Nyrox** [comment]
Thanks for the great review. I was definitely a bit sloppy yesterday (Monday 🤮).
I merged main again if you want to give that a quick review as well. :)

File: tests/sqlparser_postgres.rs
Comment (iffyio) [reply tests/sqlparser_postgres.rs]: Yeah EOF seems like natural/reasonable thanks!

**Review by iffyio — APPROVED** [review]
LGTM! Thanks for tackling this @Nyrox!
cc @alamb

**Comment by alamb** [comment]
I started looking over this PR -- and it is looking great. I took the liberty of merging up from main and fixing clippy/fmt.

I am going to test out the implications of this change by making a PR to DataFusion and test out the documentation.

**Review by alamb — APPROVED** [review]
Thank you so much Thank you so much @Nyrox and @iffyio  and @lovasoa  -- this is epic work.

Also, thank you to @lustefaniak @yuyang-ok for your comments

Many people have tried this feature but non have prevailed. 👏 . If we are ever colocated I totally owe you an in person meet up / 🍻   with a beverage of your choice

In terms of next steps:
1. I will file a ticket with the current state of the project / spans which can hopefully let us spread out the work for adding span information to the rest of parse tree over time. 
2. I think it would be good to consider changing the offsets in Location to be `u32` rather than `usize` which would reduce the memory requirements significantly I thin.
3.  I have a few ideas to improve the documentation, but I will propose some follow on PRs to do that. 

BTW I tried updating DataFusion to use this change here https://github.com/apache/datafusion/pull/13546 and it went quite smoothly

Let's leave this PR open for a few more days to get any more feedback and then plan to merge it in 🚀

**Comment by alamb** [comment]
I plan to merge this PR tomorrow unless there are any other comments

**Comment by Dandandan** [comment]
Did we run some benchmarks (e.g. cargo bench)?

**Comment by alamb** [comment]
> Did we run some benchmarks (e.g. cargo bench)?

No, I did not (it is not clear to me we have such a thing). Let me look

**Comment by alamb** [comment]
```shell
cd sqlparser_bench
cargo bench
git remote add Nyrox https://github.com/Nyrox/sqlparser-rs.git
git fetch Nyrox
git checkout Nyrox/main
```


Here is the benchmark result (it appears to be about 10%-15% slower according to the benchmark):

```
sqlparser-rs parsing benchmark/sqlparser::select
                        time:   [3.1255 µs 3.1265 µs 3.1276 µs]
                        change: [+15.235% +17.353% +19.853%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high severe
sqlparser-rs parsing benchmark/sqlparser::with_select
                        time:   [21.672 µs 21.683 µs 21.694 µs]
                        change: [+11.446% +11.549% +11.653%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe
```

**Comment by alamb** [comment]
> Thanks, looks great. I think 15% degradation is fully worth it (and might be gained back if someone looks at optimizing sqlparser-rs) :)

Yeah, I was looking at the flamegraph and there are a bunch of obvious things to improve performance (like changing `next_token` not to copy each token)

**Comment by alamb** [comment]
I filed a ticket to discuss improving performance: 
- https://github.com/apache/datafusion-sqlparser-rs/issues/1557

**Comment by Nyrox** [comment]
I also noticed that there is a bunch of calls to `<Location as Display>::fmt` which seems quite strange to me. Not sure what the test is doing, but I don't think the parser should be calling that on a non-error path? :p

**Comment by alamb** [comment]
> I also noticed that there is a bunch of calls to <Location as Display>::fmt which seems quite strange to me. Not sure what the test is doing, but I don't think the parser should be calling that on a non-error path? :p

I agree -- it is strange that Parser::expected woud show up so much. One place I see it used is generating errors 🤔 

https://github.com/apache/datafusion-sqlparser-rs/blob/2e90e105a74bf9f50f2bad6c22992759ddb06880/src/parser/mod.rs#L3424-L3429

Screenshot 2024-11-26 at 9 51 11 AM

**Comment by alamb** [comment]
Amusingly when I ran with the fixed benchmarks the result is basically the same (15% slower)

```
sqlparser-rs parsing benchmark/sqlparser::select
                        time:   [6.5723 µs 6.5856 µs 6.6010 µs]
                        change: [+14.669% +15.020% +15.364%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 12 outliers among 100 measurements (12.00%)
  11 (11.00%) high mild
  1 (1.00%) high severe
sqlparser-rs parsing benchmark/sqlparser::with_select
                        time:   [32.234 µs 32.253 µs 32.277 µs]
                        change: [+14.270% +14.402% +14.538%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 10 outliers among 100 measurements (10.00%)
  5 (5.00%) high mild
  5 (5.00%) high severe
```

[fixed-flamegraph](https://github.com/user-attachments/assets/7ceae3a0-dec2-462d-9a00-0b49afd9d2bc)

fixed-flamegraph

**Comment by alamb** [comment]
I am more convinced than ever that we could make a huge performance improvement by doing something like:
-  https://github.com/apache/datafusion-sqlparser-rs/issues/1558

**Comment by alamb** [comment]
🚀  -- thanks again @Nyrox @iffyio @lovasoa @lustefaniak @mkarbo @Dandandan  and @yuyang-ok  for helping push this along. It is pretty amazing we'll finally get new things

cc @ankrgyl  it's finally happening

**Comment by ankrgyl** [comment]
Amazing!!!

**Comment by alamb** [comment]
BTW if anyone has time to help review another PR, this one adds a bunch of documentation and examples for this feature:
- https://github.com/apache/datafusion-sqlparser-rs/pull/1549

**Comment by alamb** [comment]
BTW here is a PR from @davisp that recovers all the performance lost adding tokens (and then some) ❤️ 
- https://github.com/apache/datafusion-sqlparser-rs/pull/1587

**Pull request merged**
